### ISSUE TITLE
[BUG][Fuzzer][wrong-code] `~` on a `bool` buffer silently returns all-True instead of the logically-negated bool

### ISSUE THREAD

**Issue by Hughshine** [tile-ai/tilelang#2976]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)
- [x] I have tried the latest version of TileLang.

### What version of TileLang are you using?

0.1.13

### System information

TileLang 0.1.13 / bundled TVM fork `tile-ai/tvm@8df8ebd6` · PyTorch 2.13.0+cu130 · CUDA 13.0 · Python 3.13 · GPU: NVIDIA L40S (sm_89). Root cause is in an architecture-independent code path (frontend gate + generic C codegen), so the wrong result is expected on any target.

### Problem description

The bitwise-not operator `~` on a `bool` buffer silently returns `True` for **every** element, regardless of input. A kernel that stores `B[i] = ~A[i]` over `A = [True, False, True, False]` produces `[True, True, True, True]` instead of the expected `[False, True, False, True]`. No error, no warning — the kernel compiles and runs to completion with a deterministically wrong result.

The same `~` operator is correct on every integer width — `int8`/`int32`/`uint8` all match NumPy/PyTorch exactly (verified below) — so the defect is specific to the `bool` operand of `~`. The other bitwise operators (`&`, `|`, `^`) on `bool` are also correct; only `~` miscompiles. The computed value is genuinely wrong, not just its bool read-back: `~bool` cast to `int32` returns `[-2, -1, -2, -1]` instead of `[0, 1, 0, 1]` (see below).

Not a regression — see Provenance.

<details>
<summary>Console: bool wrong, int8/int32 controls correct</summary>

```
# bool
input : [True, False, True, False]
got   : [True, True, True, True]
expect: [False, True, False, True]      # numpy/torch: (~torch.tensor([...bool...]))

# int32 control  -> correct
got   : [-2, -1, -2, -1]   expect [-2, -1, -2, -1]   PASS
# int8 control   -> correct
got   : [-2, -1, -6, 0]    expect [-2, -1, -6, 0]    PASS  (matches numpy ~int8)
```
</details>

### Reproducible example code

```python
import tilelang
import tilelang.language as T
import torch

# --- bug: ~ on a bool buffer ---
@T.prim_func
def bool_not(A: T.Tensor((4,), "bool"), B: T.Tensor((4,), "bool")):
    with T.Kernel(1, threads=1):
        for i in T.serial(4):
            B[i] = ~A[i]

# --- control: ~ on int32 (identical kernel, int dtype) ---
@T.prim_func
def int_not(A: T.Tensor((4,), "int32"), B: T.Tensor((4,), "int32")):
    with T.Kernel(1, threads=1):
        for i in T.serial(4):
            B[i] = ~A[i]

a_bool = torch.tensor([True, False, True, False], dtype=torch.bool, device="cuda")
out_bool = tilelang.compile(bool_not, out_idx=[1])(a_bool)
print("bool  ~:", out_bool.tolist())      # -> [True, True, True, True]   (WRONG)
print("expected:", (~a_bool).tolist())    # -> [False, True, False, True] (torch reference)

a_int = torch.tensor([1, 0, 1, 0], dtype=torch.int32, device="cuda")
out_int = tilelang.compile(int_not, out_idx=[1])(a_int)
print("int32 ~:", out_int.tolist())       # -> [-2, -1, -2, -1]           (correct)
```

### Traceback

No traceback — the kernel compiles and runs to completion; the result is silently wrong and deterministic.

### Expected behavior

`~` on a `bool` buffer should return the logically-negated bool (`~True → False`, `~False → True`), matching NumPy (`np.invert`) and PyTorch (`~torch.bool_tensor`), both of which return `[False, True, False, True]` for this input. The same logical negation is already computable in TileLang today via a comparison — `B[i] = A[i] == False` returns `[False, True, False, True]` correctly — so the value is not intractable, only the `~` lowering drops it.

Note this is not merely a mismatch with numpy/torch — it is inconsistent with TileLang's *own* type system: `bool` is a 1-bit type (`DataType::Bool().bits() == 1`, and the codegen keys on `t.bits() == 1`). A bitwise-NOT of a 1-bit value is `~0→1`, `~1→0` — exactly logical negation, always in `{0,1}`. Emitting a bare C `~` cannot honor that, because C's integer-promotion rules widen a `bool` operand of `~` to a 4-byte `int` before complementing (confirmed by compiling the emitted expression: `sizeof(~(bool)true) == 4`, and `~true → -2`, `~false → -1`), so the result is always an out-of-range `int`, never a bool. The codegen therefore has to special-case a `bool` operand — emit logical negation (`!x`) or renormalize (`(~x) & 1`) — rather than defer to C's `~`, which is structurally incapable of producing a bool here. Alternatively, if `~` on `bool` is considered out of scope, the frontend could reject it the way it already rejects `~` on float; what it should not do is silently return values (`-1`/`-2`, all-`True`) outside `{True, False}`.

### Additional context

**Root cause.** The `~` frontend gate admits `bool` operands but the codegen emits an unmasked C `~`, whose result is never renormalized back into `{0, 1}` — so both `~false` and `~true` come out non-zero and read back as `True`. Concretely, [`ExprOp.__invert__`](https://github.com/tile-ai/tvm/blob/8df8ebd61659505dd7005f4820b030e983e2c005/python/tvm/tirx/expr.py#L157-L160) rejects only *float* operands (`_dtype_is_float`, which checks only [`DataTypeCode.FLOAT`](https://github.com/tile-ai/tvm/blob/8df8ebd61659505dd7005f4820b030e983e2c005/python/tvm/tirx/expr.py#L62-L64)) and passes everything else — including `bool` — to `bitwise_not`; the C codegen at [`codegen_c.cc` `bitwise_not`](https://github.com/tile-ai/tvm/blob/8df8ebd61659505dd7005f4820b030e983e2c005/src/target/source/codegen_c.cc#L677-L681) then emits a bare `(~expr)` with no per-dtype handling.

<details>
<summary>Emitted CUDA + why every element becomes <code>True</code></summary>

The generated kernel body is:

```c
B[i] = ((signed char)(~((bool)A[i])));
```

In C/C++, `~` integer-promotes its `bool` operand to `int`, so `~false → ~0 → -1` and `~true → ~1 → -2`. Both are non-zero, so when the byte is read back as a `bool`, both are `True`:

| input | `~(bool)x` | stored byte | read as bool | correct? |
|---|---|---|---|---|
| `false` | `-1` | `0xFF` | `True` | accidentally right |
| `true`  | `-2` | `0xFE` | `True` | **wrong** (should be `False`) |

Integer dtypes escape this because the truncated low bits are the correct value and are read back as an integer, not as a truthiness test — which is why `int8`/`int32` are correct.

The `-1`/`-2` is not just a read-back artifact — the computed value is itself wrong, and it leaks into any downstream integer use. `C[i] = T.Cast("int32", ~A[i])` returns `[-2, -1, -2, -1]` where `~bool` cast to int should be `[0, 1, 0, 1]` (torch). So both halves of the result are wrong; the all-`True` read-back merely hides it (both `-1` and `-2` collapse to `True`).
</details>

<details>
<summary>Scope: only <code>~</code> is affected; <code>&amp;</code> <code>|</code> <code>^</code> on bool are correct</summary>

The defect is specific to `~` — it is the only bitwise operator that produces bits above bit 0 from a `bool` operand. The others stay within `{0,1}` and are correct on the same inputs (verified this session, `A=[T,F,T,F]`, `B=[T,T,F,F]`):

| op on bool | result | correct? |
|---|---|---|
| `A & B` | `[T,F,F,F]` | yes |
| `A \| B` | `[T,T,T,F]` | yes |
| `A ^ B` | `[F,T,T,F]` | yes |
| `~A` | `[T,T,T,T]` | **no** |

`~` on integer dtypes (`uint8`/`int8`/`int32`) is also correct — only the `bool` operand of `~` miscompiles.
</details>

**Suggested fix.** For a `bool` (1-bit) operand, `~` needs to produce the negated boolean rather than a raw bitwise complement — e.g. emit `!x` (or `(x) ^ true`), or normalize the result to `{0,1}`. If instead `bool` is meant to be out of scope for `~`, the cleaner place is the `__invert__` gate, which already special-cases float — extending it to reject `bool` would surface the misuse instead of miscomputing. Which direction is right depends on whether `~` on `bool` is intended to be supported; NumPy/PyTorch both define it as logical negation.

**Provenance.** The `__invert__` float-only gate and the generic `bitwise_not` codegen are both present at the TVM fork commit pinned by tilelang 0.1.13 (`8df8ebd6`); reproduced at runtime on 0.1.13 this session (bool `[T,F,T,F]` → `[T,T,T,T]`, int8/int32 controls correct). Not a recent regression — origin before this pin not traced.

**Dedup.** No filed issue covers scalar `~`/`bitwise_not` on `bool`. The nearest bool-typed item is the reduce-op crash (`T.reduce_max`/`reduce_min`/`reduce_bitand` on a `bool` buffer, #2568) — a different path (the reduce init-value builder) and a different symptom (internal assert / crash), not this operator's silent integer-promotion.

**Reach.** The trigger is the intersection of a common operator and a rare operand type: `~` is a core builtin, but the miscompile fires only when its operand is a `bool` buffer inside device code (every integer width is correct). Grepping `examples/` and `testing/` this session: `bool` tensors do appear in examples (8 files, e.g. the deepseek sparse-MLA kernels and blocksparse GEMM), and `~`/`bitwise_not` do appear — but every `~`-on-bool occurrence is in *host-side* reference/mask code (`masked_fill(~mask, …)`, numpy `~m0 & m1`), and the one `bitwise_not` in tests is a torch host reference on ints. No kernel (device) `~` on a `bool` buffer exists in the shipped tree, which is why CI stays green.

**Impact.** The trigger is narrow — `~` must be applied to a `bool` operand in device code, and no shipped kernel does this today (every integer width is correct). When it does fire it is silent and deterministic, and the computed value is genuinely wrong rather than just a bool read-back artifact: `~bool` cast to int yields `-1`/`-2` instead of `0`/`1`, so it corrupts both boolean masks (all elements read back `True`) and any integer arithmetic derived from them.
